### PR TITLE
Sdc4x

### PR DESCRIPTION
**Pull request by torbacz**
## Description

- Implement new SCD4x Binding

## Motivation and Context

- New SCD4x Binding

## Screenshots
image

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

## Summary by CodeRabbit

- **New Features**
	- Introduced new categories: "thermometer" and "hygrometer" for enhanced device classification.
	- Added versioning configuration to streamline project releases and dependency management.
- **Documentation**
	- Established a structured manifest for managing project dependencies, improving reliability and consistency.
	- Implemented a lock file to ensure consistent library versions across environments.

### CODE REVIEW

File: devices/Scd4x/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# SCD4x - Temperature & Humidity & CO2 Sensor
+The SCD4x is Sensirion�s next generation miniature CO2 sensor. This sensor builds on the photoacoustic sensing principle and Sensirion�s patented PAsens� and CMOSens� technology to offer high accuracy at an unmatched price and smallest form factor.
```
Comment (Ellerbach) [inline devices/Scd4x/README.md:2]:
```suggestion

The SCD4x is Sensirion's next generation miniature CO2 sensor. This sensor builds on the photoacoustic sensing principle and Sensirion's patented PAsens(r) and CMOSens(r) technology to offer high accuracy at an unmatched price and smallest form factor.
```

File: devices/Scd4x/Scd4x.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+using UnitsNet;
+
+namespace Iot.Device.Scd4x
+{
+    /// <summary>
+    /// Humidity and Temperature Sensor Scd4x.
+    /// </summary>
+    public class Scd4x : IDisposable
+    {
+        private const byte CRC8POLYNOMIAL = 0x31;
```
Comment (Ellerbach) [inline devices/Scd4x/Scd4x.cs:16]: we tend to use pascal case for the constants as well

**Review by Ellerbach — COMMENTED** [review]
Great addition. Couple of remarks but all up, super quality already!

File: devices/Scd4x/Scd4x.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            }
+
+            var buffer = new byte[9];
+            result = _i2CDevice.Read(buffer);
+            if (result.Status != I2cTransferStatus.FullTransfer)
+            {
+                throw new Exception();
+            }
+
+            var serial1 = (ushort)((buffer[0] << 8) | buffer[1]);
```
[the remark this answers is not in the document]
Comment (torbacz) [reply devices/Scd4x/Scd4x.cs:64]: Can you point which one is it?

File: devices/Scd4x/Scd4x.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        }
+
+        /// <summary>
+        /// Retrieves the serial number from the SDC device.
+        /// </summary>
+        /// <returns>Serial number.</returns>
+        public string GetSerialNumber()
+        {
+            var result = _i2CDevice.Write(new byte[] { 0x36, 0x82 });
+            if (result.Status != I2cTransferStatus.FullTransfer)
```
[the remark this answers is not in the document]
Comment (torbacz) [reply devices/Scd4x/Scd4x.cs]: From my test with this device, it's working fine all the time. I'll keep that in mind when testing device. If there will be any communication issue, I'll implement retry logic :)

File: devices/Scd4x/Scd4x.cs (hunk 2 above)
Comment (torbacz) [reply devices/Scd4x/Scd4x.cs]: done

File: devices/Scd4x/Scd4x.cs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        /// <param name="offset">New offset.</param>
+        public void SetTemperatureOffset(Temperature offset)
+        {
+            var asTicks = (offset.DegreesCelsius * 65536.0 / 175.0) + 0.5f;
+            var tOffset = (ushort)asTicks;
+            var crc = GenerateCRC(tOffset);
+            var buffer = new byte[5];
+            buffer[0] = 0x24;
+            buffer[1] = 0x1D;
+            buffer[2] = (byte)((tOffset & 0xFF00) >> 8);
```
[the remark this answers is not in the document]
Comment (torbacz) [reply devices/Scd4x/Scd4x.cs]: done

File: devices/Scd4x/Scd4x.cs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        }
+
+        /// <summary>
+        /// Gets the status of data.
+        /// </summary>
+        /// <returns>True if data is ready to read.</returns>
+        public bool IsDataReady()
+        {
+            var result = _i2CDevice.Write(new byte[] { 0xE4, 0xB8 });
+            if (result.Status != I2cTransferStatus.FullTransfer)
```
[the remark this answers is not in the document]
Comment (torbacz) [reply devices/Scd4x/Scd4x.cs]: done

File: devices/Scd4x/Scd4x.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ellerbach) [reply devices/Scd4x/Scd4x.cs:58]: https://github.com/nanoframework/nanoFramework.IoT.Device/tree/develop/devices/System.Buffers.Binary.BinaryPrimitives

File: devices/Scd4x/Scd4x.cs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                    }
+                }
+            }
+
+            return crc;
+        }
+
+        private static byte GenerateCRC(ushort data)
+        {
+            var highByte = (byte)((data >> 8) & 0xFF);
```
Comment (Ellerbach) [inline devices/Scd4x/Scd4x.cs:187]: minimal as well, this is enough: `var highByte = (byte)(data >> 8);`

File: devices/Scd4x/category.txt [hunk 8]
```diff
@@ -0,0 +1,2 @@
+thermometer
+hygrometer
```
Comment (Ellerbach) [inline devices/Scd4x/category.txt:2]: and you can add `gas`

File: devices/Scd4x/samples/Properties/AssemblyInfo.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+using System.Reflection;
```
Comment (Ellerbach) [inline devices/Scd4x/samples/Properties/AssemblyInfo.cs:1]: missing the header here

**Review by Ellerbach — COMMENTED** [review]
Almost there! Thanks a lot

**Comment by torbacz** [comment]
@Ellerbach 
I think i fixed everything. Also tested with latest code - still working ;) 
image

**Review by Ellerbach — APPROVED** [review]
Amazing! Thanks a lot! And with the number of contributions, it gets easier and easier to review as well. Well done!

**Pull request merged**